packages/pds/src/db/util.ts

```diff
@@ -19,22 +21,87 @@ export const actorWhereClause = (actor: string) => {
 
 export const countAll = sql<number>`count(*)`
 
-export const paginate = <QB extends SelectQueryBuilder<any, any, any>>(
+export const paginate = <QB extends SelectQueryBuilder<any, any, any>, T>(
   qb: QB,
-  opts: {
-    limit?: number
-    before?: string
-    by: DbRef
-    secondaryOrder?: DbRef
-  },
+  opts:
+    | {
+        limit?: number
+        before?: string
+        by: DbRef
+      }
+    | {
+        limit?: number
+        before?: string
+        keyset: Keyset<T>
+      },
 ) => {
+  if ('by' in opts) {
+    return qb
+      .if(opts.limit !== undefined, (q) => q.limit(opts.limit as number))
+      .orderBy(opts.by, 'desc')
+      .if(opts.before !== undefined, (q) => q.where(opts.by, '<', opts.before))
+  }
+  const cursor = opts.keyset.unpack(opts.before)
+  const keysetCondition = getKeysetCondition(cursor, opts.keyset)
   return qb
-    .orderBy(opts.by, 'desc')
-    .if(opts.secondaryOrder !== undefined, (q) =>
-      q.orderBy(opts.secondaryOrder as DbRef, 'desc'),
-    )
     .if(opts.limit !== undefined, (q) => q.limit(opts.limit as number))
-    .if(opts.before !== undefined, (q) => q.where(opts.by, '<', opts.before))
+    .orderBy(opts.keyset.primary, 'desc')
+    .orderBy(opts.keyset.secondary, 'desc')
+    .if(keysetCondition !== undefined, (qb) =>
+      keysetCondition ? qb.where(keysetCondition) : qb,
+    )
+}
+
+type DefaultRow = { createdAt: string; uri: string }
+export abstract class Keyset<T = DefaultRow> {
```

Comment:
could these just be params on the constructor instead of abstract values?

Reply:
Yeah, let's do it.  I think this ended-up a little funky because of some different things I was trying out along the way.

Reply:
nice. i like how this cleaned up 👍